kani-driver/src/concrete_playback.rs

```diff
@@ -18,42 +21,64 @@ use std::process::Command;
 
 impl KaniSession {
     /// The main driver for generating concrete playback unit tests and adding them to source code.
-    pub fn gen_and_add_concrete_playback(&self, output_filename: &Path, harness: &HarnessMetadata) {
+    pub fn gen_and_add_concrete_playback(
+        &self,
+        harness: &HarnessMetadata,
+        verification_output: &VerificationOutput,
+    ) -> Result<()> {
         if self.args.concrete_playback.is_none() {
-            return;
+            return Ok(());
         }
 
-        assert!(
+        ensure!(
             self.args.output_format != crate::args::OutputFormat::Old,
             "The Kani argument `--output-format old` is not supported with the concrete playback feature."
         );
 
-        let concrete_vals = parser::extract_concrete_vals(output_filename).expect(
-            "Something went wrong when trying to get concrete values from the CBMC output file",
-        );
-        let concrete_playback = format_unit_test(&harness.mangled_name, &concrete_vals);
+        if verification_output.status == VerificationStatus::Success {
+            if !self.args.quiet {
+                println!(
+                    "INFO: The concrete playback feature does not generate unit tests when verification succeeds."
+                )
+            }
+            return Ok(());
+        }
 
-        if let Some(playback_mode) = &self.args.concrete_playback && *playback_mode == ConcretePlaybackMode::Print && !self.args.quiet {
-            println!(
-                "Concrete playback unit test for `{}`:\n```\n{}\n```",
-                &harness.mangled_name, &concrete_playback.unit_test_str
-            );
-            println!(
-                "To automatically add the concrete playback unit test `{}` to the src code, run Kani with `--concrete-playback=InPlace`.",
-                &concrete_playback.unit_test_name
+        if let (Some(processed_items), Some(playback_mode)) =
+            (&verification_output.processed_items, &self.args.concrete_playback)
+        {
+            let concrete_vals = extract_from_processed_items(processed_items).expect(
+                "Something went wrong when trying to get concrete values from the CBMC output",
             );
-        }
+            let concrete_playback = format_unit_test(&harness.mangled_name, &concrete_vals);
 
-        if let Some(playback_mode) = &self.args.concrete_playback && *playback_mode == ConcretePlaybackMode::InPlace {
-            if !self.args.quiet {
+            if *playback_mode == ConcretePlaybackMode::Print && !self.args.quiet {
```

Comment:
Not for this PR, but shouldn't this be an error? Users shouldn't be able to select `print` mode and `--quiet` at the same time.

Reply:
Nice catch! I added an `ensure` for this.